File: frontend/src/api/friendDashboard.ts
import { apiClient } from "./client";
import type { DashboardAlbum } from "./dashboard";

export interface FriendDashboardEntry {
  album: DashboardAlbum;
  mutual_date: string;
  similarity_users: number | null;
  similarity_a_vs_spotify: number | null;
  similarity_b_vs_spotify: number | null;
  spotify_top5_indices: number[];
  user_a_top_track_indices: number[];
  user_b_top_track_indices: number[];
  mean_score: number;
  user_a_score: number;
  user_b_score: number;
}

export interface FriendDashboardResponse {
  // null for an ad-hoc comparison between two non-friends (computed live).
  friendship_id: number | null;
  user_a_username: string;
  user_b_username: string;
  entries: FriendDashboardEntry[];
}

/** Where a pair comparison's data comes from: a friendship (precomputed) or an
 *  arbitrary viewable user (computed live). Drives the fetch, the persisted-state
 *  namespace, and the per-album detail link. */
export type ComparisonSource =
  | { kind: "friendship"; friendshipId: number }
  | { kind: "user"; username: string };

export function fetchComparison(
  source: ComparisonSource
): Promise<FriendDashboardResponse> {
  return source.kind === "friendship"
    ? getFriendDashboard(source.friendshipId)
    : getUserComparison(source.username);
}

export async function getFriendDashboard(
  friendshipId: number
): Promise<FriendDashboardResponse> {
  const { data } = await apiClient.get<FriendDashboardResponse>(
    `/friendships/${friendshipId}/dashboard`
  );
  return data;
}

/** Live pair comparison with any viewable (non-friend) user. Same shape as the
 *  friend dashboard, but computed on the fly server-side (friendship_id === null). */
export async function getUserComparison(
  username: string
): Promise<FriendDashboardResponse> {
  const { data } = await apiClient.get<FriendDashboardResponse>(
    `/users/${encodeURIComponent(username)}/comparison`
  );
  return data;
}

File: backend/app/schemas/friend_dashboard.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.dashboard import DashboardAlbum


class FriendDashboardEntryOut(BaseModel):
    album: DashboardAlbum
    mutual_date: datetime
    similarity_users: float | None
    similarity_a_vs_spotify: float | None
    similarity_b_vs_spotify: float | None
    spotify_top5_indices: list[int]
    user_a_top_track_indices: list[int]
    user_b_top_track_indices: list[int]
    mean_score: float
    user_a_score: float
    user_b_score: float


class FriendDashboardResponse(BaseModel):
    # Present for a friend comparison (precomputed); None for an ad-hoc comparison
    # between two arbitrary viewable users (computed live, no friendship row).
    friendship_id: int | None
    user_a_username: str
    user_b_username: str
    # Profile picture URLs for the two compared users (None when unset). Used by
    # the comparison page's similarity tiles to show each user's avatar.
    user_a_picture_url: str | None
    user_b_picture_url: str | None
    entries: list[FriendDashboardEntryOut]
